### PR TITLE
Consistent Rot3 printing

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR unifies the printing functionality of Rot3 to use the `Eigen::IOFormat` formatting, thus ensuring consistency.

The aim would be to help the user visually identify more easily what the rotation matrices are, as well as copy them into another language such as python.

I've also
1. updated the printing of `Pose3`, `NavState` and `Rot3` to provide default spacing when using the custom prefix string
2. removed extra newlines by removing newlines in the `<<` overload
3. removed `endl` in `<<` so as to avoid unnecessary buffer flushes.

For the below snippet,

```cpp
Pose3 p;
cout << p << endl;
p.print();
```

the previous format looked like this:

```sh
# result of cout
|1, 0, 0|
|0, 1, 0|
|0, 0, 1|

[0, 0, 0]';

# result of .print()
R:
[
	1, 0, 0;
 	0, 1, 0;
 	0, 0, 1
]
[0, 0, 0]';

```

The new format would give the output:

```sh
# result of cout
R: [
	1, 0, 0;
	0, 1, 0;
	0, 0, 1
]
t: [0, 0, 0]'
# result of .print()
R: [
	1, 0, 0;
	0, 1, 0;
	0, 0, 1
]
t: [0, 0, 0]'
```

---
This change is [Reviewable](https://reviewable.io/reviews/borglab/gtsam/358)

### CODE REVIEW

**Comment by varunagrawal** [comment]
Additionally, this is how `Pose3` now prints

```
R: [
	1, 0, 0;
	0, 1, 0;
	0, 0, 1
]
t: [0, 0, 0]'
```

and this is for `NavState`

```
R: [
	1, 0, 0;
	0, 1, 0;
	0, 0, 1
]
p: [0, 0, 0]'
v: [0, 0, 0]'
```

**Comment by dellaert** [comment]
Ask other people for reviews ?

**Comment by dellaert** [comment]
> Ask other people for reviews ?

Dude. I don’t mean ask everyone at once :-)

File: gtsam/base/Matrix.h [hunk 1]
```diff
@@ -76,6 +76,18 @@ GTSAM_MAKE_MATRIX_DEFS(9);
...
+  Eigen::StreamPrecision, // precision
+  Eigen::DontAlignCols, // flags set such that rowSpacers are not added
+  ", ", // coeffSeparator
+  ";\n", // rowSeparator
+  "\t",  // rowPrefix
+  "", // rowSuffix
+  "[\n", // matPrefix
+  "\n]" // matSuffix
+);
+
```
Comment (jlblancoc) [inline gtsam/base/Matrix.h:90]: Hmmm... I'm not comfortable with this symbol defined here in the header. It was safer as it was before, as a static var inside a smaller scope :-) 
My understanding (and experience) is that it *might* lead to link errors in some compilers/build types. 

Why not defining it as a static variable in the corresponding .cpp file, and make this a getter?

```
// In the .h

// Matrix formatting arguments when printing. Akin to Matlab style.
const Eigen::IOFormat &matlabFormat(); 
```

```
// in the .cpp

static const Eigen::IOFormat matlab(
      Eigen::StreamPrecision, // precision
      0, // flags
      ", ", // coeffSeparator
      ";\n", // rowSeparator
      "\t",  // rowPrefix
      "", // rowSuffix
      "[\n", // matPrefix
      "\n]" // matSuffix
      );

const Eigen::IOFormat &gtsam::matlabFormat() { return matlab; }
```

File: gtsam/geometry/Rot3.cpp [hunk 2]
```diff
@@ -32,7 +32,8 @@ namespace gtsam {
 
 /* ************************************************************************* */
 void Rot3::print(const std::string& s) const {
-  gtsam::print((Matrix)matrix(), s);
+  cout << (s.empty() ? "R: " : s + " ");
+  gtsam::print((Matrix)matrix());
```
Comment (jlblancoc) [inline gtsam/geometry/Rot3.cpp:36]: prefer `static_cast<>()` instead of old C cast?

File: gtsam/navigation/NavState.cpp [hunk 3]
```diff
@@ -88,15 +88,15 @@ Matrix7 NavState::matrix() const {
 
 //------------------------------------------------------------------------------
 ostream& operator<<(ostream& os, const NavState& state) {
-  os << "R:" << state.attitude();
-  os << "p:" << state.position() << endl;
-  os << "v:" << Point3(state.velocity()) << endl;
+  os << "R: " << state.attitude() << "\n";
+  os << "p: " << state.position() << "\n";
```
Comment (jlblancoc) [inline gtsam/navigation/NavState.cpp:92]: :+1:

File: gtsam/base/Matrix.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/base/Matrix.h:90]: That's an excellent point I had not considered, and the proposed solution is elegant! I'll get on it.

File: gtsam/geometry/Rot3.cpp (hunk 2 above)
Comment (varunagrawal) [reply gtsam/geometry/Rot3.cpp:36]: Yes! I was trying not to introduce too many changes, but this is a very good change to introduce.

**Comment by varunagrawal** [comment]
@dellaert sorry about that. I figured casting a wide net would yield faster results. :sweat_smile:

**Comment by varunagrawal** [comment]
@dellaert @jlblancoc now that 4.0.3 is out, I would really like to see this PR merged in. :slightly_smiling_face:

**Comment by varunagrawal** [comment]
Whoops. Yeah the other PR is based off this one hence you're seeing the issue repeat itself. Give me a few quick minutes to update this.

**Comment by varunagrawal** [comment]
So I am a bit confused about the `endl` to `\n` conversion left here. From my understanding, we want to do a buffer flush when calling `print()` and that is the only place where we use `endl`. :-)

**Comment by jlblancoc** [comment]
> So I am a bit confused about the `endl` to `\n` conversion left here. From my understanding, we want to do a buffer flush when calling `print()` and that is the only place where we use `endl`. :-)

Hmm... I'm thinking of, say, we call print() on a Values or an entire FactorGraph. One flush at the end is ok... but 1000s of them, after printing (recursively) each value/matrix/vector/factor/etc. seems not necessary :-) 

Still, it's probably a minor point, so feel free of leaving it as you think it's better.

**Comment by varunagrawal** [comment]
I completely agree with your point. The issue is that the Values/FactorGraph print method just calls the internal print method of the various values/factors, so we need to add the `endl` over there. I think that should be left to a future PR?

**Comment by jlblancoc** [comment]
Ok. I don't want to stop you doing anything more productive! 
Merge it when you want to.

**Comment by varunagrawal** [comment]
Merging because Travis keeps timing out and we will be moving to Github Actions.

**Pull request merged**